cupy/cuda/cupy_cub.cu

```diff
@@ -97,6 +101,25 @@ __host__ __device__ __forceinline__ complex<double> Max::operator()(const comple
     else {return max(a, b);}
 }
 
+#if __CUDACC_VER_MAJOR__ >= 9 && (__CUDA_ARCH__ >= 530 || !defined(__CUDA_ARCH__))
+// specialization for half for handling NaNs
+template <>
+__host__ __device__ __forceinline__ __half Max::operator()(const __half &a, const __half &b) const
+{
+    // NumPy behavior: NaN is always chosen!
+#ifdef __CUDA_ARCH__
+    if (__hisnan(a)) {return a;}
+    else if (__hisnan(b)) {return b;}
+    else {return CUB_MAX(a, b);}
+#else
+    // TODO: avoid cast to float
+    if (isnan(__half2float(a))) {return a;}
+    else if (isnan(__half2float(b))) {return b;}
+    else {return __float2half(CUB_MAX(__half2float(a), __half2float(b)));}
+#endif
```

Comment:
@y1r I think it's a bit too cumbersome. If I read `cuda_fp16.h` right, the struct `__half` is defined on both host and device. Could you define a `isnan` for `__half` like this
```C
__host__ __device__ __forceinline__ bool isnan(const __half& x) {
  #ifdef __CUDA_ARCH__
      return __hisnan(x);
  #else
      // TODO: avoid cast to float
      return isnan(__half2float(x));
  #endif
}
```

and reuse it everywhere to reduce repeats? That `TODO` could be a separate PR by examining the bit fields. (I don't think it's worth the effort though, since CUB kernels only needs the `__device__` path and it's there only to pass compiler.)